node/network/availability-distribution/src/lib.rs

```diff
@@ -360,18 +365,28 @@ where
 			.map(|(peer, _view)| peer.clone())
 			.collect();
 
+		// If we have no peers that are interested, skip the rest.
+		if peers.is_empty() {
+			continue;
+		}
+
 		let per_candidate = state.per_candidate.entry(candidate_hash).or_default();
 
 		let validator_count = per_candidate.validators.len();
 
 		// distribute all erasure messages to interested peers
 		for chunk_index in 0u32..(validator_count as u32) {
-			let erasure_chunk = if let Some(message) = per_candidate.message_vault.get(&chunk_index) {
-				message.erasure_chunk.clone()
-			} else if let Some(erasure_chunk) = query_chunk(ctx, candidate_hash, chunk_index as ValidatorIndex).await? {
+			let erasure_chunk = if let Some(erasure_chunk) =
+				query_chunk(ctx, candidate_hash, chunk_index as ValidatorIndex).await? {
```

Comment:
Why go to the availability store unconditionally, if the message may exist in the message vault?

Reply:
This change makes the vault obsolete, and these two changes are co-dependant.

Reply:
At this point, the message can not yet exist in the message vault.

If we are switching to a new view, it means we either fetched old candidates which would be handled by `on_peer_view_changed`. Or we fetched new candidates. If we fetched new candidates, there can also be no message in the message vault, because we would have rejected any message up to now for this candidate, because our view wasn't yet aware of the relay block that introduced this candidate.

Reply:
I don't believe this is true.

Note that this is doing `state.cached_live_candidates_unioned`, not `query_pending_availability`. If the parablock was pending availability in a recent ancestor or a sibling block that we've already imported, then we'll have already queried the data from disk.

Reply:
Yeah sorry. I see what you mean. When peer updates it view before us, we would not send it the required data.

Reply:
Actually that is what I meant with the other comment but not this one.

Let's say our initial view is `[A]`. `live_candidates(A) = [X, Y]`. Now we get a new block B and our view is `[A, B]`. `live_candidates(B) = [Y, Z]`. So `cached_live_candidates_unioned = [X, Y, Z]`. We already have a `per_candidate` with filled out message vault for `[X, Y]`, so we don't need to go to the av-store again. We only need to go to disk for the new `Z` candidate.